### PR TITLE
fix(swap): approval tx of amounts with lots decimals

### LINKED ISSUE #21325: Not enough DAI allowance given to tokenTransferProxy error is shown in case of sending tokens using max value button
### Problem:
The issue is that after entering a value with 18 decimals (e.g., 2.84042568174435125), the approved value is only 15 decimals (e.g., 2.840425681744351).

image

This discrepancy, as shown in [Optimistic Etherscan](https://optimistic.etherscan.io/), means that the user is attempting to swap a value (2.84042568174435125) that is not fully approved.

### Steps to Reproduce:
1. Restore a user with a token that has 18 decimal places.
2. Go to the swap page.
3. Tap the "max" value.
4. Approve the max value.
5. Try to confirm the transaction.

### Actual Result:
The following error is displayed: "Not enough DAI allowance given to tokenTransferProxy."

https://github.com/user-attachments/assets/ac5fb29c-482f-4efa-b8c9-abc9f8adf5dd

### Expected Result:

Potential solutions:

The 18 decimals token should be able to be approved. Current approval works correct on desktop

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### Logs
[Status-debug-logs (2).zip](https://github.com/user-attachments/files/17134945/Status-debug-logs.2.zip)

### PR DESCRIPTION
**Pull request by briansztamfater**
fixes #21325 

### Summary

This PR fixes a rounding issue when converting hex numbers with lots of decimals, causing issues in some swap transactions (check issue for more context).

Tested these changes locally and were able to execute both approval and swap transactions without any issues:

https://optimistic.etherscan.io/tx/0x8c028e3dde799d43cfb589ae9360120ec54548f48e37fdf0f39d3eedc2f27feb
https://optimistic.etherscan.io/tx/0xd4e82302434dcaa96591c079835c4d08131ed92998d4bfe162224c229b4670ce

#### Platforms

- Android
- iOS

##### Functional

- wallet / transactions

### Steps to test

1. Restore a user with a token that has 18 decimal places.
2. Go to the swap page.
3. Tap the "max" value.
4. Approve the max value.
5. Verify the approval amount was correct on the blockchain explorer
6. Try to confirm the transaction.
7. Verify the transaction is executed correctly

status: ready

### CODE REVIEW

File: src/utils/money.cljs [hunk 1]
```diff
@@ -151,6 +152,11 @@
   [^js bn]
   (str "0x" (to-string bn 16)))
 
+(defn from-hex
+  [hex-str]
+  (-> (gint/fromString hex-str 16)
```
Comment (briansztamfater) [inline src/utils/money.cljs]: We use `goog.math.Integer` for converting hex to decimal because JavaScript’s native number type has precision limitations for large integers (above 2^53). `goog.math.Integer` provides arbitrary-precision arithmetic, allowing accurate handling of large values without rounding errors.

**Review by mohsen-ghafouri — APPROVED** [review]
👍

**Review by shivekkhurana — APPROVED** [review]

File: src/utils/money.cljs (hunk 1 above)
Comment (clauxx) [reply src/utils/money.cljs]: Maybe this belongs in the `utils.hex` ns

Comment (vkjr) [reply src/utils/money.cljs]: @briansztamfater, can't we use the bignumber itself? The [documentation](https://mikemcl.github.io/bignumber.js/) says it can construct from hex string.
excerpt from doc:
```js
new BigNumber('ff.8', 16)                  // '255.5'
new BigNumber('0xff.8')                    // '255.5'
```

Also, it would be great to add a few tests to cover edge cases. Recently we experienced [exceptions](https://github.com/status-im/status-mobile/pull/21282) thrown by bignumber and it would be nice to avoid this.

**Comment by VolodLytvynenko** [comment]
Hi @briansztamfater thank you for PR. No issues from my side. PR is ready to be merged

File: src/utils/money.cljs (hunk 1 above)
Comment (briansztamfater) [reply src/utils/money.cljs]: > Maybe this belongs in the `utils.hex` ns

@clauxx there are lots of hex-related functions on through different namespaces of the app. We need to define how we want to handle this. For now, as this is a function to create a BigNumber object based on a hex string, I put this on `utils.money` namespace, because that is the namespace where bignumber functions are. I feel we need to revisit all these namespaces, probably we need to move bignumber-related functions to a `utils.bignumber` namespace, but I didn't want to do this in this PR just to fix the bug. I will create an issue so we can do a proper refactor!

> @briansztamfater, can't we use the bignumber itself? The [documentation](https://mikemcl.github.io/bignumber.js/) says it can construct from hex string. excerpt from doc:
> 
> ```js
> new BigNumber('ff.8', 16)                  // '255.5'
> new BigNumber('0xff.8')                    // '255.5'
> ```
> 
> Also, it would be great to add a few tests to cover edge cases. Recently we experienced [exceptions](https://github.com/status-im/status-mobile/pull/21282) thrown by bignumber and it would be nice to avoid this.

@vkjr I didn't know BigNumber had a constructor that admits hex strings. That's much cleaner indeed. I added some tests, and changed the approach to use the BigNumber constructor.

**Comment by briansztamfater** [comment]
@VolodLytvynenko thanks for testing! Based on the feedback, I modified the approach. Can you please verify it didn't break anything? Thanks again!

**Comment by VolodLytvynenko** [comment]
Hi @briansztamfater thank you for update. PR is rechecked. No issues from my side. Ready to be merged

**Review by smohamedjavid — APPROVED** [review]
